File: LeetCode/22_Generate_Parentheses.js
var generateParenthesis = function(n) {
    var res = []
    
    if (n < 1) return res;
    var generate = function (res, str, ll, rr) {
        if (ll || rr) {
            if (rr > ll) generate(res, str + ')', ll, rr - 1)
            if (ll) generate(res, str + '(', ll - 1, rr);
        } else {
            res.push(str)
        }
    }
    generate(res, '', n, n);
    return res;
};


// Time Complexity : O( (4^n)/(n^(-1)) )?
// Space Complexity: O( (4^n)/(n^(-1)) )?

File: LeetCode/22_Generate_Parentheses.py

def generateParenthesis(n):
    res = []

    def dfs(l, r, temp):
        if not l and not r:
            res.append(temp[:])
            print(res)
        if l:
            dfs(l - 1, r, temp + "(")
        if l < r:
            dfs(l, r - 1, temp + ")")
    dfs(n, n, "")
    # return res
    print(res)





generateParenthesis(2)